### ISSUE TITLE
lost password/secret questions

### ISSUE THREAD

**Issue by Kindred-999** [SimpleMachines/SMF#4829]
using this option triggered the following errors

(build as of 1000 on 10-Jul)

```
Warning
: mysqli_stmt_bind_param() expects parameter 1 to be mysqli_stmt, boolean given in
/.../Sources/Subs-Db-mysql.php
on line
994


Warning
: mysqli_stmt_execute() expects parameter 1 to be mysqli_stmt, boolean given in
/.../Sources/Subs-Db-mysql.php
on line
995


Warning
: mysqli_stmt_bind_param() expects parameter 1 to be mysqli_stmt, boolean given in
/.../Sources/Subs-Db-mysql.php
on line
994


Warning
: mysqli_stmt_execute() expects parameter 1 to be mysqli_stmt, boolean given in
/.../Sources/Subs-Db-mysql.php
on line
995
```

**Comment by albertlast** [comment]
This error is typical for a outdated table schema,
specialy in this cases that you log_error table is not on the newest version:
https://github.com/SimpleMachines/SMF2.1/blob/a3d3e2ad2322c64337f4245106906a3a16b72df4/other/install_2-1_mysql.sql#L334-L350
In your case the backtrace col should be missing.

**Comment by Kindred-999** [comment]
hmmm.... I ran the upgrade.php - shouldn't that have corrected any missing tables/schema?


correction. Just checked. That column IS present. Upgrade did correct it.

export of my log_errors table
```
CREATE TABLE `smf_log_errors` (
  `id_error` mediumint(8) UNSIGNED NOT NULL,
  `log_time` int(10) UNSIGNED NOT NULL DEFAULT '0',
  `id_member` mediumint(8) UNSIGNED NOT NULL DEFAULT '0',
  `ip` varbinary(16) DEFAULT NULL,
  `url` text NOT NULL,
  `message` text NOT NULL,
  `session` varchar(128) NOT NULL DEFAULT '                                                                ',
  `error_type` char(15) NOT NULL DEFAULT 'general',
  `file` varchar(255) NOT NULL DEFAULT '',
  `line` mediumint(8) UNSIGNED NOT NULL DEFAULT '0',
  `backtrace` varchar(10000) NOT NULL DEFAULT ''
) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```

**Comment by albertlast** [comment]
So was the table before already correct or now after the running of the upgrade?

**Comment by wintstar** [comment]
Depends on which pull request he uses.

There was something  ...
https://github.com/SimpleMachines/SMF2.1/pull/4762

**Comment by Kindred-999** [comment]
I don't know if it was correct before. I believe not (the site was taken down months ago and just restored)
The build I used was from 2 days ago.

**Comment by wintstar** [comment]
Did you make the changes in Settings.php?

From Pull request #4777

https://github.com/SimpleMachines/SMF2.1/commit/61d52300437ad78e42dcb6b121e86ce68c588375#diff-686b483befc40ea4318cc8ea6676892cR118

**Comment by jdarwood007** [comment]
@Kindred-999 

Can you still reproduce it?  I was able to upgrade just fine and didn't get this.

**Comment by Kindred-999** [comment]
Added the line to Settings.php and it appears to have solved the error.

Do note, this was an upgrade from a previous (beta3/4) installation.   So, the question arises - without someone pointing out the change to Settings.php - how would this have expected to have been fixed/avoided?

I ran upgrade.php, which successfully updated the database tables.
The one thing that is NOT overwritten from upgrades is Settings.php -- so we definitely need to call out in a notification during the upgrade, if any changes were made to Settings.php

**Comment by illori** [comment]
that is why upgrading from beta to beta is not supported

**Comment by wintstar** [comment]
One must follow the development intensively to upgrade from beta to beta. There are still changes in Settings.php from Beta3 to Beta 4. When SMF 2.1 is released, I think the upgrade procedure will also change a bit. But it is not yet so far.

**Comment by jdarwood007** [comment]
Upgrading from a beta version has nothing to do with this.  The PR that added this didn't add any support in the upgrade process to ensure that variable exists or to handle it properly when it doesn't.

Right now any forum that upgrades from the previous version would break.  Including upgrading from 2.0, 1.1 or 1.0.

The fix here is to change the logic to handle when its not set somewhere in the database startup.  At the very least.

**Comment by jdarwood007** [comment]
Kindred,

Looking into this more.

The code should allow the setting to be missing from Settings.php.  Can you confirm if removing the changes from Settings.php breaks it again?

What error specifically are you getting.  The same ones from before or new ones?

**Comment by Kindred-999** [comment]
I'll look into it on Sunday. I'm heading out for the whole day tomorrow.

**Comment by Sesquipedalian** [comment]
Can this be closed?

**Comment by sbulen** [comment]
I cannot reproduce this exact issue.   The lost password/secret question function seems to work.  

The presence or absence of the mb4 setting doesn't make a difference.  (I performed a 2.0=>2.1 upgrade with the 'migrate to a new settings file' checkbox _un_checked, so the setting is not present in settings.php.)

I found a separate issue - if you make an error (e.g., passwords don't match or don't meet criteria when setting your new password) you get in a session timed out loop & have to start over.  Not sure if I should log that as a new issue or not.
